### PR TITLE
Pass correct MP Table entry count to unblock spec-compliant guests

### LINKED ISSUE #3045: MP Table has incorrect entry count
# Describe the bug

The MP Configuration Table Header (per Intel Multiprocessing Specification) includes fields "BASE TABLE LENGTH" and "ENTRY COUNT".  The former specifies the length of the table (header plus entries), while the latter specifies the number of entries.  The specification states that "Software must step through each entry in the base table until it reaches ENTRY COUNT".

Firecracker sets ENTRY COUNT to zero.  Linux boots only because of a kernel bug -- it ignores ENTRY COUNT and instead scans through counting bytes until it hits BASE TABLE LENGTH.

## To Reproduce

Try to boot an OS which obeys the Intel Multiprocessing Specification (e.g. FreeBSD).

## Expected behaviour

ENTRY COUNT should be initialized properly to the number of table entries.

## Environment

`[Author TODO: Please supply the following information):]`
`[ - Firecracker version.]`
`[ - Host and guest kernel versions.]`
`[ - Rootfs used.]`
`[ - Architecture.]`
`[ - Any other relevant software versions.]`

## Additional context

`[Author TODO: How has this bug affected you?]`

`[Author TODO: What are you trying to achieve?]`

`[Author TODO: Do you have any idea of what the solution might be?]`

## Checks

- [ ] Have you searched the Firecracker Issues database for similar problems?
- [ ] Have you read the existing relevant Firecracker documentation?
- [ ] Are you certain the bug being reported is a Firecracker issue?

### PR DESCRIPTION
**Pull request by acj**
Resolves #3045.

## Changes

This PR counts the number of entries added to the MultiProcessor Configuration Table and passes it to the guest through the `oemcount` field on `mpspec::mpc_table`.

## Reason

Guest operating systems that comply with the [Intel MP spec](https://pdos.csail.mit.edu/6.828/2018/readings/ia32/MPspec.pdf) (see Section 4), such as FreeBSD, fail to boot because the Entry Count field in the MP Configuration Table header is unset (defaults to 0). When Firecracker sets that field to a reasonable value, those guests are able to discover the MP configuration and continue booting.

## Notes

I've tested this on a FreeBSD 14 guest with the patches from the `feature/pvh` branch and a few kernel `printf`s added to verify that the `MPTABLE_LINUX_BUG_COMPAT` option wasn't interfering (it was enabled but didn't need to calculate the entry count because Firecracker provided one). I also successfully booted a few Linux guests, both with and without the PVH patches.

I added a unit test but am unsure whether any integration tests need to be added/updated. If so, I'm happy to do that.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [X] If a specific issue led to this PR, this PR closes the issue.
- [X] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] API changes follow the [Runbook for Firecracker API changes][2].
- [ ] User-facing changes are mentioned in `CHANGELOG.md`.
- [X] All added/changed functionality is tested.
- [ ] New `TODO`s link to an issue.
- [X] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [?] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: src/vmm/src/arch/x86_64/mptable.rs [hunk 1]
```diff
@@ -369,6 +378,24 @@ mod tests {
...
+        )
+        .unwrap();
+
+        setup_mptable(&mem, num_cpus).unwrap();
+
+        let mpf_intel: mpspec::mpf_intel = mem.read_obj(GuestAddress(MPTABLE_START)).unwrap();
+        let mpc_offset = GuestAddress(u64::from(mpf_intel.physptr));
+        let mpc_table: mpspec::mpc_table = mem.read_obj(mpc_offset).unwrap();
+
+        assert_eq!(mpc_table.oemcount, 30);
```
Comment (bchalios) [inline src/vmm/src/arch/x86_64/mptable.rs:396]: nit: could you add a comment here with the reason why this adds up to 30?

**Review by bchalios — COMMENTED** [review]
Thanks a lot for this! LGTM, just a small nit.

File: src/vmm/src/arch/x86_64/mptable.rs [hunk 2]
```diff
@@ -369,6 +378,39 @@ mod tests {
...
+            // Intel floating point
+            + 1
+            // CPU
+            + num_cpus as u16
+            // IOAPIC
+            + 1
+            // ISA Bus
+            + 1
+            // IRQ
+            + (IRQ_MAX as u16) + 1
```
Comment (acj) [inline src/vmm/src/arch/x86_64/mptable.rs:406]: Extra +1 is from ranging over `0..=IRQ_MAX` (inclusive) here:

https://github.com/firecracker-microvm/firecracker/blob/1afbacadfa7e9891e654928044ecfb322642c179/src/vmm/src/arch/x86_64/mptable.rs#L211

**Comment by acj** [comment]
Fixed the clippy warnings in the test code

**Review by roypat — APPROVED** [review]

**Review by bchalios — APPROVED** [review]
LGTM! Thanks a lot for this @acj

**Pull request merged**